### PR TITLE
feature/Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by NitBravoA92**
# 🚩Milestone 8: Integration specs for views and fix n+1 problems | add-integration-test-and-fix-n+1-problems - >>branch<< -

### Hi there! @NitBravoA92 and @CesarHerr  here. 👋

This is the PR for the **Blog app - Integration specs for views and fix n+1 problems** activity. The main objective of this milestone is to write integration tests for all of the views used in your project using Capybara gem and Secure app from n+1 problems.

### Here is a summary of what has been done:

- [x] Install and setup the `capybara` and  `webdrivers` gems

- [x] Refactor the `index` action of the `Posts` controller to fix an n+1 problem caused by too many unnecessary queries to the database

- [x] Create all the test cases for the index page of users: 
        - Test the username of all other users.
        - Test the profile picture for each user.
        - Test the number of posts each user has written.
        - Test when the user clicks on a username, it's redirected to that user's show page.

- [x] Create all the test cases for the show page of users: 
        - Test the user's profile picture.
        - Test the user's username.
        - Test the number of posts the user has written.
        - Test the user's bio.
        - Test the user's latest 3 posts.
        - Test a button that lets you view all of a user's posts.
        - Test when the user clicks a user's post, it redirects to that post's show page.
        - Test when the user clicks to see all posts, it redirects to the user's post's index page.

- [x] Create all the test cases for the index page of posts: 
        - Test the user's profile picture.
        - Test the user's username.
        - Test the number of posts the user has written.
        - Test a post's title.
        - Test a post's body.
        - Test the first comments on a post.
        - Test how many comments a post has.
        - Test how many likes a post has.
        - Test a section for pagination if there are more posts than fit on the view.
        - Test when a user clicks on a post, it redirects to that post's show page.
        - Test when a user clicks on the 'Create a post' button, it redirects to the post's new page.

- [x] Create all the test cases for the show page of posts: 
        - Test the post's title.
        - Test who wrote the post.
        - Test how many comments it has.
        - Test how many likes it has.
        - Test the post body.
        - Test the username of each commentor.
        - Test the comment each commentor left.

- [x] Create all the test cases for the new page of posts: 
        - Test the page's title.
        - Test the form labels: 'Title' and 'Text'
        - Test when the user clicks on the 'Save post' button, it saves a new Post on the database

- [x] Create all the test cases for the new page of comments: 
        - Test the page's title.
        - Test the form label: 'Text'
        - Test when the user clicks on the 'Save comment' button, it saves a new Comment for a post on the database

- [x] Update the README.md file

## Screenshots with the results of the n+1 problems fixed
 
- **BEFORE - index action of PostsController**:
posts_index_action_before

- **AFTER - index action of PostsController**:
posts_index_action_after

## ⭐To the code reviewer 👨‍💻

💝 We really thank 🙇‍♂️ you, dear code reviewer 👨‍🎨 for dedicating your precious 🥇 time ⌚ to check the PR 🧾 .

We kindly ask 🙏 dear code reviewer 🤵 that if there is any issue 🦯 in this project, please do list 📃 them in a descriptive 💡 manner and give your best suggestions 🎁 if needed.

If you think some significant issues need to be changed ♻️ please kindly contact 📞 us through Zoom or Slack, We are available on Microverse UTC-6 time zone.

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    Post.create(author: @user, title: 'Post 3', text: 'This is the content of Post 3')
+
+    visit user_path(@user)
+
+    expect(page).to have_css('.user_posts .post_item_container', count: 3)
+
+    expect(page).to have_content 'Post 1'
+    expect(page).to have_content 'Post 2'
+    expect(page).to have_content 'Post 3'
+  end
```
Comment (FrederickMih) [inline spec/features/user_show_spec.rb]: - Hi @NitBravoA92, I don't think your expected result resolved the hypothesis raised in the "it" block. We should be testing that if a user created more than 4 posts, only the 3 most recent posts should display on this page.

File: spec/features/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    post1 = Post.create(author: @user, title: 'Post 1', text: 'This is the content of Post 1')
+
+    visit user_posts_path(@user)
+
+    click_link 'Post 1'
+
+    expect(page).to have_current_path(user_post_path(@user, post1))
+  end
+
+  it "Clicking on the 'Create a post' button should redirect to post's new page" do
```
Comment (FrederickMih) [inline spec/features/post_index_spec.rb:89]: - Awesome job testing additional scenario here :rocket:

**Review by FrederickMih — APPROVED** [review]
## Hi :raising_hand: :raising_hand: @, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 

:tada: :champagne:  :tada: :champagne: 

## Highlights :green_circle: 

- Fantastic job including additional tests for user's input fields for both posts and comments :rocket: 
- Unit tests runs with success :+1: 
- PR is professional :+1: 

## Optional suggestions
**N/A**

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb (hunk 1 above)
Comment (NitBravoA92) [reply spec/features/user_show_spec.rb]: @FrederickMih Thank you very much for the excellent feedback you have given me. I have refactored this test to correctly validate that only the most recent posts from a user are displayed on this page

File: spec/features/post_index_spec.rb (hunk 2 above)
Comment (NitBravoA92) [reply spec/features/post_index_spec.rb:89]: @FrederickMih Thank you very much for highlighting the additional test cases.

**Pull request merged**